### PR TITLE
Add derived columns.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: autoprecompiles/src/constraint_optimizer.rs [hunk 1]
```diff
@@ -120,6 +120,12 @@ pub fn trivial_simplifications<P: FieldElement, V: Ord + Clone + Eq + Hash + Dis
     let constraint_system = remove_redundant_constraints(constraint_system);
     stats_logger.log("removing redundant constraints", &constraint_system);
 
+    let constraint_system = remove_unreferenced_derived_variables(constraint_system);
```
Comment (Schaeff) [inline autoprecompiles/src/constraint_optimizer.rs:123]: Maybe if it's cheap we should run this within the other optimizers instead of as a separate optimizer, so that `constraint_system` is always a "well formed" constraint system, as in, with no unreferenced derived variable.

File: autoprecompiles/src/lib.rs [hunk 2]
```diff
@@ -163,12 +163,40 @@ pub enum BusInteractionKind {
     Receive,
 }
 
+/// A machine comprised of algebraic constraints, bus interactions and potentially derived columns.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct SymbolicMachine<T> {
+    /// Constraints whose expressions have to evaluate to zero for an assignment to be satisfying.
     pub constraints: Vec<SymbolicConstraint<T>>,
+    /// A bus interaction that models communication with other machines / chips or static lookups.
```
Comment (Schaeff) [inline autoprecompiles/src/lib.rs]:
```suggestion
    /// Bus interactions that models communication with other machines / chips or static lookups.
```

File: autoprecompiles/src/lib.rs [hunk 3]
```diff
@@ -163,12 +163,40 @@ pub enum BusInteractionKind {
...
+    /// to compute their values from other columns.
+    pub derived_columns: Vec<(AlgebraicReference, ComputationMethod<T>)>,
 }
 
+/// A machine comprised of algebraic constraints, bus interactions and potentially derived columns.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SymbolicMachineOld<T> {
+    /// Constraints whose expressions have to evaluate to zero for an assignment to be satisfying.
+    pub constraints: Vec<SymbolicConstraint<T>>,
+    /// A bus interaction that models communication with other machines / chips or static lookups.
```
Comment (Schaeff) [inline autoprecompiles/src/lib.rs]:
```suggestion
    /// Bus interactions that model communication with other machines / chips or static lookups.
```

File: autoprecompiles/src/symbolic_machine_generator.rs [hunk 4]
```diff
@@ -26,6 +27,21 @@ pub fn convert_machine_field_type<T, U>(
...
+            .map(|(v, method)| {
+                let method = match method {
+                    ComputationMethod::Constant(c) => {
+                        ComputationMethod::Constant(convert_field_element(c))
+                    }
+                    ComputationMethod::InverseOrZero(e) => ComputationMethod::InverseOrZero(
+                        convert_expression(e, convert_field_element),
+                    ),
+                };
+                (v.clone(), method)
```
Comment (Schaeff) [inline autoprecompiles/src/symbolic_machine_generator.rs]: Is cloning required here?

File: autoprecompiles/src/lib.rs [hunk 5]
```diff
@@ -163,12 +163,40 @@ pub enum BusInteractionKind {
     Receive,
 }
 
+/// A machine comprised of algebraic constraints, bus interactions and potentially derived columns.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct SymbolicMachine<T> {
+    /// Constraints whose expressions have to evaluate to zero for an assignment to be satisfying.
     pub constraints: Vec<SymbolicConstraint<T>>,
+    /// A bus interaction that models communication with other machines / chips or static lookups.
     pub bus_interactions: Vec<SymbolicBusInteraction<T>>,
```
Comment (Schaeff) [inline autoprecompiles/src/lib.rs:172]: How about wrapping the old struct here? It would make sense that air extraction returns the old one, then apc creation takes all the old ones and returns a new one (with derived columns). This is also nice for the client, where we can use the derived columns during execution, and during constraint evaluation continue using the old struct without the derived columns.

Comment (Schaeff) [reply autoprecompiles/src/lib.rs:172]: Maybe the derived columns could be added to the Apc struct instead of to the symbolic machine.

File: constraint-solver/src/indexed_constraint_system.rs [hunk 6]
```diff
@@ -90,34 +93,63 @@ impl<T, V> Default for IndexedConstraintSystemWithQueue<T, V> {
...
+/// A reference to an item in the constraint system, based on the index.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
 enum ConstraintSystemItem {
+    /// A reference to an algebraic constraint.
     AlgebraicConstraint(usize),
+    /// A reference to a bus interaction.
     BusInteraction(usize),
+    /// A reference to a derived variable. This is only used internal to the
+    /// IndexedConstraintSystem.
+    DerivedVariable(usize),
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs:105]: Love these

File: constraint-solver/src/indexed_constraint_system.rs [hunk 7]
```diff
@@ -90,34 +93,63 @@ impl<T, V> Default for IndexedConstraintSystemWithQueue<T, V> {
...
 
-/// A reference to an item in the constraint system.
+/// A reference to an item in the constraint system, based on the index.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
 enum ConstraintSystemItem {
+    /// A reference to an algebraic constraint.
     AlgebraicConstraint(usize),
+    /// A reference to a bus interaction.
     BusInteraction(usize),
+    /// A reference to a derived variable. This is only used internal to the
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs]:
```suggestion
    /// A reference to a derived variable. This is only used internally to the
```

File: constraint-solver/src/indexed_constraint_system.rs [hunk 8]
```diff
@@ -90,34 +93,63 @@ impl<T, V> Default for IndexedConstraintSystemWithQueue<T, V> {
...
+    /// of items.
+    fn index(&self) -> usize {
+        match self {
+            ConstraintSystemItem::AlgebraicConstraint(index)
+            | ConstraintSystemItem::BusInteraction(index)
+            | ConstraintSystemItem::DerivedVariable(index) => *index,
+        }
+    }
+
+    /// Returns true if this constraint system item is an actual constraint and not a derived variable.
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs]: `is_constraint` sounds like algebraic constraint to me. Maybe we can use `!is_derived_vairable`? No strong opinion though.

File: constraint-solver/src/indexed_constraint_system.rs [hunk 9]
```diff
@@ -392,8 +441,21 @@ fn variable_occurrences<T: RuntimeConstant, V: Hash + Eq + Clone>(
                 .unique()
                 .map(move |v| (v.clone(), ConstraintSystemItem::BusInteraction(i)))
         });
+    let occurrences_in_derived_variables = constraint_system
+        .derived_variables
+        .iter()
+        .enumerate()
+        // We ignore the derived variables itself because it is not a constraint
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs]:
```suggestion
        // We ignore the derived variable itself because it is not a constraint
```

File: constraint-solver/src/indexed_constraint_system.rs [hunk 10]
```diff
@@ -392,8 +441,21 @@ fn variable_occurrences<T: RuntimeConstant, V: Hash + Eq + Clone>(
...
+        // and does not matter in substitutions (if we substitute the derived
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs:443]: Can you explain this? If we have `a := 1/b` we do not register `a` as being accessed in this item. Why?

File: constraint-solver/src/indexed_constraint_system.rs [hunk 11]
```diff
@@ -415,6 +477,13 @@ fn substitute_by_known_in_item<T: RuntimeConstant + Substitutable<V>, V: Ord + C
                 .fields_mut()
                 .for_each(|expr| expr.substitute_by_known(variable, substitution));
         }
+        ConstraintSystemItem::DerivedVariable(i) => match &mut constraint_system.derived_variables
+            [i]
+            .1
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs]: Let's have a struct for this?

File: constraint-solver/src/indexed_constraint_system.rs [hunk 12]
```diff
@@ -435,6 +504,13 @@ fn substitute_by_unknown_in_item<T: RuntimeConstant + Substitutable<V>, V: Ord +
                 .fields_mut()
                 .for_each(|expr| expr.substitute_by_unknown(variable, substitution));
         }
+        ConstraintSystemItem::DerivedVariable(i) => match &mut constraint_system.derived_variables
+            [i]
+            .1
+        {
+            ComputationMethod::InverseOrZero(e) => e.substitute_by_unknown(variable, substitution),
+            ComputationMethod::Constant(_) => {}
+        },
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs]: Can we implement `substitute_by_{unknown, known}` on `ComputationMethod`?

File: constraint-solver/src/indexed_constraint_system.rs [hunk 13]
```diff
@@ -471,9 +547,11 @@ where
 
     /// Removes the next item from the queue and returns it.
     pub fn pop_front<'a>(&'a mut self) -> Option<ConstraintRef<'a, T, V>> {
-        self.queue
-            .pop_front()
-            .map(|item| item.to_constraint_ref(&self.constraint_system.constraint_system))
+        self.queue.pop_front().map(|item| {
+            item.try_to_constraint_ref(&self.constraint_system.constraint_system)
+                // Derived variables should never be in the queue.
+                .unwrap()
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs:547]: I'm starting to question the value of having the derived variables in the same pool as constraints and bus interactions, since we have a to treat it as an edge case in a lot of places. What's the tradeoff here?

File: constraint-solver/src/indexed_constraint_system.rs [hunk 14]
```diff
@@ -743,4 +824,38 @@ mod tests {
             .to_string();
         assert_eq!(items_with_z, "x - z = 0");
     }
+
+    #[test]
+    fn substitute_in_deried_columns() {
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs]:
```suggestion
    fn substitute_in_derived_columns() {
```

File: constraint-solver/src/indexed_constraint_system.rs [hunk 15]
```diff
@@ -743,4 +824,38 @@ mod tests {
...
+        let mut system: IndexedConstraintSystem<_, _> = ConstraintSystem::<
+            GoldilocksField,
+            &'static str,
```
Comment (Schaeff) [inline constraint-solver/src/indexed_constraint_system.rs]: inferred?

File: openvm/src/powdr_extension/executor/mod.rs [hunk 16]
```diff
@@ -170,6 +172,7 @@ impl<F: PrimeField32> PowdrExecutor<F> {
             &self.air_by_opcode_id,
             self.number_of_calls,
             &self.apc,
+            F::ONE, // TODO this is stupid but I did not find a way to get the 1 in the field.
```
Comment (Schaeff) [inline openvm/src/powdr_extension/executor/mod.rs:175]: This is fine for now.

File: autoprecompiles/src/constraint_optimizer.rs (hunk 1 above)
Comment (chriseth) [reply autoprecompiles/src/constraint_optimizer.rs:123]: We could, but I don't really see the benefit in having no unreferenced derived variables. It is relatively cheap, but also not free (it needs to iterate through the whole system once).

File: autoprecompiles/src/lib.rs (hunk 5 above)
Comment (chriseth) [reply autoprecompiles/src/lib.rs:172]: I'm not sure I understand. What do you mean by "old struct"?

File: constraint-solver/src/indexed_constraint_system.rs (hunk 10 above)
Comment (chriseth) [reply constraint-solver/src/indexed_constraint_system.rs:443]: Yes. The occurrence map is used for substitutions: If we substitute `b` by `y` we also want to do that inside the computation method. But if we substitute `a` by `t`, we don't really care, we also don't want to substitute `t := 1/b` because `t` might not be a derived variable (if it is, it has its own computation method). Also if we want to figure out the referenced unused variables, we don't want to count `a` in here.

File: constraint-solver/src/indexed_constraint_system.rs (hunk 13 above)
Comment (chriseth) [reply constraint-solver/src/indexed_constraint_system.rs:547]: Indeed, they are different, but most of the operations we apply on constraints we also have to apply on the derived variables. We can pull them out, but then the place where we have to treat them as edge cases will just move. I could try to remove them from the queue, but then I would need to introduce a new data structure just for that.

File: openvm/src/powdr_extension/executor/mod.rs (hunk 16 above)
Comment (chriseth) [reply openvm/src/powdr_extension/executor/mod.rs:175]: What do you mean?

File: autoprecompiles/src/lib.rs (hunk 5 above)
Comment (Schaeff) [reply autoprecompiles/src/lib.rs:172]: I mean `SymbolicMachineOld`. My point is that in this one is still relevant in some places (like constraint evaluation), as we know there won't be derived variables, so maybe the derived variables should be put somewhere else.

File: autoprecompiles/src/lib.rs [hunk 17]
```diff
@@ -163,12 +163,40 @@ pub enum BusInteractionKind {
...
+    /// Constraints whose expressions have to evaluate to zero for an assignment to be satisfying.
     pub constraints: Vec<SymbolicConstraint<T>>,
+    /// Bus interactions that model communication with other machines / chips or static lookups.
     pub bus_interactions: Vec<SymbolicBusInteraction<T>>,
+    /// Columns that have been newly created during the optimization process with a method
+    /// to compute their values from other columns.
+    pub derived_columns: Vec<(AlgebraicReference, ComputationMethod<T>)>,
 }
 
+/// A machine comprised of algebraic constraints, bus interactions and potentially derived columns.
```
Comment (Schaeff) [inline autoprecompiles/src/lib.rs:178]:
```suggestion
/// A machine comprised of algebraic constraints and bus interactions
```

File: constraint-solver/src/indexed_constraint_system.rs (hunk 13 above)
Comment (Schaeff) [reply constraint-solver/src/indexed_constraint_system.rs:547]: I would personally be more in favour of a new data structure, but won't block if you think it's better like this.

File: openvm/src/powdr_extension/executor/mod.rs (hunk 16 above)
Comment (Schaeff) [reply openvm/src/powdr_extension/executor/mod.rs:175]: I don't think this is stupid, maybe we can refactor tracegen to not require passing one like this eventually.

**Review by Schaeff — APPROVED** [review]
Some remaining non-blocking comments

File: autoprecompiles/src/lib.rs (hunk 5 above)
Comment (chriseth) [reply autoprecompiles/src/lib.rs:172]: Oh sorry, SymbolicMachineOld was just a leftover to transfer the cbor files.
